harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -122,23 +122,67 @@ def _init_device(self) -> None:
             self.device = torch.device("cpu")
         check.is_not_none(self.device)
 
-    def _init_model_and_optimizer(self) -> None:
+    def _init_model_and_optimizers(self) -> None:
+        # backwards compatibility: remove this when removing the old functions.
+        if util.is_overridden(self.trial._build_optimizers, PyTorchTrial) or util.is_overridden(
+            self.trial._build_lr_schedulers, PyTorchTrial
+        ):
+            check.true(util.is_overridden(self.trial._build_optimizers, PyTorchTrial))
+            check.true(util.is_overridden(self.trial._build_lr_schedulers, PyTorchTrial))
+        else:
+            logging.warning(
+                "optimizer and create_lr_scheduler are deprecated. "
+                "Please use build_optimizers and build_lr_schedulers instead."
+            )
+
         self.context.model = self.trial.build_model()
+        self.context.model = self.context.model.to(self.device)
 
         # TODO: Check that optimizer is not an amp optimizer.
-        self.context.optimizer = self.trial.optimizer(self.context.model)
+        optimizers = self.trial._build_optimizers(self.context.model)
+        if isinstance(optimizers, torch.optim.Optimizer):  # type: ignore
+            self.context.optimizers = [optimizers]
+        elif isinstance(optimizers, list):
+            check.gt_eq(len(optimizers), 1)
+            self.context.optimizers = optimizers
+        else:
+            raise det.errors.InvalidExperimentException(
+                "invalid return type from build_optimizers. "
+                "Please see an instance list or the only instance of "
+                "torch.optim.Optimizer."
+            )
 
-        self._init_device()
-        self.context.model = self.context.model.to(self.device)
+        lr_schedulers = self.trial._build_lr_schedulers(optimizers)
```

Comment:
blocking: this ordering will lead to the schedulers pointing to the wrong optimizer in the cases where we use horovod. We need to move this after the horovod optimizer initialization.

Reply:
When we use Horovod, it replaces the original optimizer with the Horovod optimizer. The `lr_scheduler` maintains a reference to the optimizer that is passed in to it in the constructor, updating its `lr`. So if we build the schedulers before we create the Horovod optimizer, they will be updating the `lr` of the wrong optimizer(s).